### PR TITLE
feat(sampling): Add async ring buffer and Spyre-optimized samplers

### PR DESCRIPTION
**Pull request by dmholtz**
## Description

This PR optimizes the Spyre sampling path in three stages to remove the main latency bottlenecks from token generation while preserving correctness.
To ease the review, each stage is a separate commit.

### Problem
The default sampling path spends too much time in the critical decoding loop on CPU-heavy work:
- Exponential noise generation is computed synchronously in the sampling path.
- Softmax normalization remains expensive for large vocabularies.
- Tensor-parallel CPU ranks redundantly sample and diverge from one another.

### Solution
1. Stage 1: Async noise pre-sampling
   - Introduce an asynchronous ring buffer that pre-generates exponential noise on a background thread.
   - Replace synchronous `exponential_()` calls in the critical path with zero-copy buffer access.
   - Add `SpyreTopKTopPSampler` and `SpyreSampler` integration for the Spyre path.

2. Stage 2: Single-rank sampling on TP
   - Only TP rank 0 performs sampling and broadcasts the selected token IDs to the remaining ranks.
   - This removes redundant CPU work and keeps all ranks aligned on the same tokens.

3. Stage 3: Log-space Gumbel sampling
   - Reformulate sampling as `argmax(logits - log_noise)` instead of `argmax(log(softmax(logits)) + log_noise)`.
   - This preserves sampling behavior while removing softmax from the critical path.

### Key Insight
The three optimizations are complementary: async noise generation removes expensive background work from the hot loop, TP rank gating removes redundant computation and divergence, and log-space Gumbel sampling removes softmax entirely. Together they reduce sampling latency without changing output distribution.
Performance tests show that OTPS increase by >30% for Granite 4.1 while staying in a fixed, thight compute budget.

## Related Issues

n/a

## Test Plan

- Unit tests to verify ring-buffer correctness and sampling correctnes
- Performance tests

## Checklist

- [x] I have read the contributing guidelines
- [x] My code follows the project's code style (`bash format.sh`)
- [x] I have added tests for my changes
- [ ] I have updated the documentation where needed
- [x] My commits include a `Signed-off-by:` line (DCO compliance)

### CODE REVIEW

File: tests/v1/sample/test_spyre_topk_topp_sampler.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (dilipgb) [inline tests/v1/sample/test_spyre_topk_topp_sampler.py:1]: nitpick: correct the file name: test_spyre_topk_topp_sampler

**Review by dilipgb — CHANGES_REQUESTED** [review]

**Comment by dilipgb** [comment]
Otherwise looks good to me. Pele tests passed fine. Maybe we wanted to run e2e tests on card once? @gkumbhat can you please take a look at this once?

File: tests/v1/sample/test_spyre_topk_topp_sampler.py (hunk 1 above)
Comment (dmholtz) [reply tests/v1/sample/test_spyre_topk_topp_sampler.py:1]: Done

**Comment by dmholtz** [comment]
Thank you for your review @dilipgb, let's run another round of tests to get this PR green

**Comment by dilipgb** [comment]
@joerunde would take look at this PR once?

**Comment by tdoublep** [comment]
Can we rebase the PR against the new version of vLLM to check there is no interaction?

**Comment by dmholtz** [comment]
@tdoublep I rebased this PR to main, it went smooth without any conflicts. Could you please re-run the tests to confirm the behavior with vLLM v0.27.1?

**Review by tdoublep — APPROVED** [review]
LGTM - thanks for the awesome work!

**Comment by tdoublep** [comment]
@dilipgb Please take a final look (you requested changes, which is blocking merge). Tests are passing against latests main

File: sendnn_inference/model_executor/model_loader/spyre.py [hunk 2]
```diff
@@ -112,8 +114,18 @@ def __init__(
         rank: int,
     ) -> None:
         super().__init__()
-
-        self.sampler = Sampler()
+        vllm_config_compatible = SpyreSampler.is_vllm_config_compatible(vllm_config)
+        if not vllm_config_compatible:
+            warnings.warn(
```
Comment (R3hankhan123) [inline sendnn_inference/model_executor/model_loader/spyre.py]: Can we use vllm's logger here instead of warnings

File: sendnn_inference/v1/sample/spyre_topk_topp_sampler.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        self,
+        logits: torch.Tensor,
+        generators: dict[int, torch.Generator],
+        k: torch.Tensor | None,
+        p: torch.Tensor | None,
+    ) -> tuple[torch.Tensor, torch.Tensor | None]:
+        """Apply top-k/top-p filtering and sample tokens using pre-drawn noise."""
+
+        if generators:
+            warnings.warn(
```
Comment (R3hankhan123) [inline sendnn_inference/v1/sample/spyre_topk_topp_sampler.py]: Can we use vllm's logger here instead of warnings

**Review by R3hankhan123 — COMMENTED** [review]
1-2 nits rest it looks good to me

File: sendnn_inference/v1/sample/spyre_topk_topp_sampler.py (hunk 3 above)
Comment (dmholtz) [reply sendnn_inference/v1/sample/spyre_topk_topp_sampler.py]: @R3hankhan123 Thank you for the suggestion. I intentionally used warnings.warn() here because this is intended to be a Python warning that callers can filter/control, rather than diagnostic logging. Does vLLM's logger support the same (to avoid spamming the logs in every forward)? Since this is a NIT, I would leave it unchanged.

File: sendnn_inference/model_executor/model_loader/spyre.py (hunk 2 above)
Comment (dmholtz) [reply sendnn_inference/model_executor/model_loader/spyre.py]: This is part of the constructor, so my argument from below does not really apply here...

Comment (dmholtz) [reply sendnn_inference/model_executor/model_loader/spyre.py]: @R3hankhan123 I pushed a commit to replace the warnings.warn call here. Please re-run the CI.

File: sendnn_inference/v1/sample/spyre_topk_topp_sampler.py (hunk 3 above)
Comment (R3hankhan123) [reply sendnn_inference/v1/sample/spyre_topk_topp_sampler.py]: yup you can use warning_once function call in vllm's logger

**Comment by dmholtz** [comment]
@R3hankhan123 Thanks for your review, I addressed both NITs.

**Review by R3hankhan123 — APPROVED** [review]
LGTM

**Review by dilipgb — APPROVED** [review]
LGTM
